File: mirea/cources/2026-2027/aii-july/01/Part-04-Binary-search/examples/example-13.cpp
// Стандартные алгоритмы бинарного поиска в C++

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int main() {
    // Чтение отсортированного массива и значения x
    int n;
    cin >> n;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
        cin >> a[i];
    }

    int x;
    cin >> x;

    // Находим границы вхождений x и используем их для ответов на несколько вопросов
    auto lower_it = lower_bound(a.begin(), a.end(), x);
    auto upper_it = upper_bound(a.begin(), a.end(), x);
    bool found = binary_search(a.begin(), a.end(), x);
    auto [equal_left_it, equal_right_it] = equal_range(a.begin(), a.end(), x);

    int lower = (int)(lower_it - a.begin());
    int upper = (int)(upper_it - a.begin());
    bool found_by_lower = lower_it != a.end() && *lower_it == x;
    int equal_left = (int)(equal_left_it - a.begin());
    int equal_right = (int)(equal_right_it - a.begin());
    int equal_count = (int)(equal_right_it - equal_left_it);
    int less = lower;
    int equal = equal_count;
    int greater = (int)a.size() - upper;

    // Проверяем, что разные стандартные функции согласованы между собой
    if (found != found_by_lower || lower != equal_left || upper != equal_right) {
        return 1;
    }

    // Выводим позиции, факт наличия элемента и количества элементов по отношению к x
    cout << lower << ' ' << upper << '\n';
    cout << (found ? "YES" : "NO") << '\n';
    cout << equal_left << ' ' << equal_right << ' ' << equal_count << '\n';
    cout << less << ' ' << equal << ' ' << greater << '\n';
}

File: mirea/cources/2026-2027/aii-july/01/Part-04-Binary-search/examples/example-13.py
# Стандартные приёмы бинарного поиска в Python3

from bisect import bisect_left, bisect_right

# Чтение отсортированного массива и значения x
n = int(input())
a = list(map(int, input().split()))
x = int(input())

# В Python3 функции bisect_left и bisect_right играют роль lower_bound и upper_bound
left = bisect_left(a, x)
right = bisect_right(a, x)
found = left < len(a) and a[left] == x

# Диапазон [left, right) является аналогом equal_range,
# а по его границам легко посчитать количества элементов
equal_left = left
equal_right = right
equal_count = right - left
less = left
equal = equal_count
greater = len(a) - right

# Выводим позиции, факт наличия элемента и количества элементов по отношению к x
print(left, right)
print("YES" if found else "NO")
print(equal_left, equal_right, equal_count)
print(less, equal, greater)